mongo/datadog_checks/mongo/mongo.py

```diff
@@ -650,6 +653,117 @@ def _collect_indexes_stats(self, instance, db, tags):
             except Exception as e:
                 self.log.error("Could not fetch indexes stats for collection %s: %s", coll_name, e)
 
+    def _get_submission_method(self, method_name):
+        if method_name not in ALLOWED_CUSTOM_METRICS_TYPES:
+            raise ValueError('Metric type {} is not one of {}.'.format(method_name, ALLOWED_CUSTOM_METRICS_TYPES))
+        return getattr(self, method_name)
+
+    @staticmethod
+    def _extract_command_from_mongo_query(mongo_query):
+        """Extract the command (find, count or aggregate) from the query. Even though mongo and pymongo are supposed
+        to work with the query as a single document, pymongo expects the command to be the `first` element of the
+        query dict.
+        Because python 2 dicts are not ordered, the command is extracted to be later run as the first argument
+        of pymongo `runcommand`
+        """
+        for command in ALLOWED_CUSTOM_QUERIES_COMMANDS:
+            if command in mongo_query:
+                return command
+        raise ValueError("Custom query command must be of type {}".format(ALLOWED_CUSTOM_QUERIES_COMMANDS))
+
+    def _collect_custom_metrics_for_query(self, db, raw_query, tags):
+        """Validates the raw_query object, executes the mongo query then submits the metrics to datadog"""
+        metric_prefix = raw_query.get('metric_prefix')
+        if not metric_prefix:  # no cov
+            raise ValueError("Custom query field `metric_prefix` is required")
+        metric_prefix = metric_prefix.rstrip('.')
+
+        mongo_query = deepcopy(raw_query.get('query'))
+        if not mongo_query:  # no cov
+            raise ValueError("Custom query field `query` is required")
+        mongo_command = self._extract_command_from_mongo_query(mongo_query)
+        collection_name = mongo_query[mongo_command]
+        del mongo_query[mongo_command]
```

Comment:
@FlorianVeaux 
Sorry for giving the feedback so late on this.

Having the command (aggregate, count, find) in the query object from the config seems not ideal. 

1) There a risk of collision between the command and the content of the query itself (for example find has many fields itself https://docs.mongodb.com/manual/reference/command/find/#dbcmd.find)
2) We could avoid having to pop/del the command (aggregate, count, find) from the query object.
3) Not mixing command and query seems a bit cleaner and might avoid confusions too.

Hum, why not just separate the `query` object and the command ? For example:

```yaml
    custom_queries:
      - metric_prefix: mongo.example1
        collection: <COLLECTION_NAME>
        command: aggregate/find/count
        query:
          pipeline: ...
          cursor: {} 
        fields: ...
```